src/algebra/order/kleene.lean

```diff
@@ -0,0 +1,291 @@
+/-
+Copyright (c) 2022 Siddhartha Prasad, Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Siddhartha Prasad, Yaël Dillies
+-/
+import algebra.order.ring.canonical
+import algebra.ring.pi
+import algebra.ring.prod
+import order.hom.complete_lattice
+
+/-!
+# Kleene Algebras
+
+This file defines idempotent semirings and Kleene algebras, which are used extensively in the theory
+of computation.
+
+An idempotent semiring is a semiring whose addition is idempotent. An idempotent semiring is
+naturally a semilattice by setting `a ≤ b` if `a + b = b`.
+
+A Kleene algebra is an idempotent semiring equipped with an additional unary operator `∗`, the
+Kleene star.
+
+## Main declarations
+
+* `idem_semiring`: Idempotent semiring
+* `idem_comm_semiring`: Idempotent commutative semiring
+* `kleene_algebra`: Idempotent commutative semiring
+
+## References
+
+* (Kozen, D. . A completeness theorem for Kleene algebras and the algebra of regular events.)
+  [https://www.cs.cornell.edu/~kozen/Papers/ka.pdf].
+* https://planetmath.org/idempotentsemiring
+* https://encyclopediaofmath.org/wiki/Idempotent_semi-ring
+* https://planetmath.org/kleene_algebra
+
+## TODO
+
+Instances for `add_opposite`, `mul_opposite`
+
+## Tags
+
+kleene algebra, idempotent semiring
+-/
+
+set_option old_structure_cmd true
+
+open function
+
+universe u
+variables {α β ι : Type*} {π : ι → Type*}
+
+/-- An idempotent semiring is a semiring with the additional property that addition is idempotent.
+-/
+@[protect_proj]
+class idem_semiring (α : Type u) extends semiring α, semilattice_sup α :=
+(sup := (+))
+(add_eq_sup : ∀ a b : α, a + b = a ⊔ b . try_refl_tac)
+(bot : α := 0)
+(bot_le : ∀ a, bot ≤ a)
```

Comment:
Why not also extend `order_bot` to get these last two fields?

Reply:
This can't be done because `order_bot` is a mixin.